Mathlib/Combinatorics/SimpleGraph/Matching.lean

```diff
@@ -130,6 +130,126 @@ theorem IsPerfectMatching.even_card {M : Subgraph G} [Fintype V] (h : M.IsPerfec
   simpa only [h.2.card_verts] using IsMatching.even_card h.1
 #align simple_graph.subgraph.is_perfect_matching.even_card SimpleGraph.Subgraph.IsPerfectMatching.even_card
 
+lemma isPerfectMatching_induce_supp_isMatching {M : Subgraph G} (h : Subgraph.IsPerfectMatching M)
+    (c : ConnectedComponent G) : (M.induce c.supp).IsMatching  := by
+    intro v hv
+    obtain ⟨ w, hw ⟩ := h.1 (h.2 _)
+    use w
+    constructor
+    · constructor
+      · assumption
+      · constructor
+        · rw [ConnectedComponent.mem_supp_iff] at *
+          rw [← hv]
+          apply ConnectedComponent.connectedComponentMk_eq_of_adj
+          apply M.adj_sub
+          rw [Subgraph.adj_comm]
+          exact hw.1
+        · exact hw.1
+    · intro y hy
+      exact hw.2 y hy.2.2
+
 end Subgraph
 
+namespace ConnectedComponent
+
+section Finite
+
+variable [Fintype V] [DecidableEq V] [DecidableRel G.Adj]
+
+/--
+ Decidable instance for membership of support of a connected component.
+-/
+instance (c : G.ConnectedComponent) (v : V) : Decidable (v ∈ c.supp) :=
+  c.recOn
+    (fun w => by simp only [ConnectedComponent.mem_supp_iff, ConnectedComponent.eq]; infer_instance)
+    (fun _ _ _ _ => Subsingleton.elim _ _)
+
+lemma even_if_perfect_matching {M : Subgraph G} (c : ConnectedComponent G)
+    (hM : Subgraph.IsPerfectMatching M) : Even (Fintype.card ↑(ConnectedComponent.supp c)) := by
+    classical
+    obtain ⟨ k , hk ⟩ := (M.isPerfectMatching_induce_supp_isMatching hM c).even_card
+    use k
+    rw [← hk, Subgraph.induce_verts, Fintype.card_ofFinset]
+    congr
+    simp only [ConnectedComponent.mem_supp_iff, Finset.mem_univ, forall_true_left]
+    exact Set.filter_mem_univ_eq_toFinset fun x => connectedComponentMk G x = c
+
+/--
+  Local instance of Fintype for sets in a Fintype.
+  Chosen as local because it is noncomputable.
+-/
+noncomputable local instance (u : Set V) : Fintype u := Fintype.ofFinite ↑u
```

Comment:
This is a pretty bad idea. You should add the necessary fintype assumptions by hand. Otherwise you will have a hard time applying your lemmas

Reply:
Gotcha, that makes sense. Removed it:)